File: app/src/main/java/org/stypox/dicio/skills/telephone/TelephoneOutput.java
package org.stypox.dicio.skills.telephone;

import static org.stypox.dicio.Sentences_en.telephone;

import android.content.ContentResolver;
import android.content.Context;
import android.util.Pair;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.core.content.res.ResourcesCompat;

import org.dicio.skill.Skill;
import org.dicio.skill.chain.OutputGenerator;
import org.dicio.skill.standard.StandardResult;
import org.stypox.dicio.R;
import org.stypox.dicio.output.graphical.GraphicalOutputUtils;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

public class TelephoneOutput extends OutputGenerator<StandardResult> {

    public void generate(final StandardResult data) {
        final ContentResolver contentResolver = ctx().android().getContentResolver();
        final String userContactName = data.getCapturingGroup(telephone.who).trim();
        final List<Contact> contacts
                = Contact.getFilteredSortedContacts(contentResolver, userContactName);

        final List<Pair<String, List<String>>> validContacts = new ArrayList<>();
        final LinearLayout output = GraphicalOutputUtils.buildVerticalLinearLayout(ctx().android(),
                ResourcesCompat.getDrawable(ctx().android().getResources(),
                        R.drawable.divider_items, null));

        for (int i = 0; validContacts.size() < 5 && i < contacts.size(); ++i) {
            final Contact contact = contacts.get(i);
            final List<String> numbers = contact.getNumbers(contentResolver);
            if (numbers.isEmpty()) {
                continue;
            }

            if (validContacts.isEmpty() // the first (i.e. lowest-distance) valid contact
                    && contact.getDistance() < 3 // a low enough distance
                    && numbers.size() == 1 // it has just one number
                    && (contacts.size() <= i + 1 // the next contact (if any) has a higher distance
                    || contacts.get(i + 1).getDistance() > contact.getDistance())) {
                // very close match with just one number and without distance ties: call it directly
                ConfirmCallOutput.callAfterConfirmation(this, contact.getName(), numbers.get(0));
                return;
            }

            validContacts.add(new Pair<>(contact.getName(), numbers));
            addNumbersToOutput(contact, numbers, output, ctx().android());
        }

        // this point will not be reached if a very close match was found
        if (validContacts.isEmpty()) {
            ctx().getSpeechOutputDevice().speak(ctx().android().getString(
                    R.string.skill_telephone_unknown_contact));
        } else {
            ctx().getSpeechOutputDevice().speak(ctx().android().getString(
                    R.string.skill_telephone_found_contacts, validContacts.size()));
            ctx().getGraphicalOutputDevice().display(output);

            if (validContacts.size() == 1 // there is exactly one valid contact and ...
                    // ... either it has exactly one number, or we would be forced to use
                    // RecognizeNameIO in any case, which only uses the first number anyway
                    && (validContacts.get(0).second.size() == 1
                    || ctx().getNumberParserFormatter() == null)) {
                // not a good enough match, but since we have only this, call it directly
                final Pair<String, List<String>> contact = validContacts.get(0);
                ConfirmCallOutput.callAfterConfirmation(this, contact.first, contact.second.get(0));

            } else {
                // ask the user which one to call (assumes items in validContacts are in the same
                // order as those displayed by addNumbersToOutput)
                final List<Skill> nextSkills = new ArrayList<>();
                final List<NameNumberPair> recognizeNameContacts = validContacts.stream()
                        // when saying the name, there is no way to distinguish between
                        // different numbers, so just use the first one
                        .map(pair -> new NameNumberPair(pair.first, pair.second.get(0)))
                        .collect(Collectors.toList());
                nextSkills.add(new ContactChooserName(recognizeNameContacts));

                if (ctx().getNumberParserFormatter() != null) {
                    final List<NameNumberPair> recognizeIndexContacts = validContacts.stream()
                            .flatMap(pair -> pair.second.stream()
                                    .map(number -> new NameNumberPair(pair.first, number)))
                            .collect(Collectors.toList());
                    nextSkills.add(new ContactChooserIndex(recognizeIndexContacts));
                }

                setNextSkills(nextSkills);
            }
        }
    }

    private void addNumbersToOutput(final Contact contact,
                                    final List<String> numbers,
                                    final LinearLayout output,
                                    final Context context) {
        for (int j = 0; j < numbers.size(); ++j) {
            final View view;
            if (j == 0) {
                view = GraphicalOutputUtils.inflate(context, R.layout.skill_telephone_contact);
                ((TextView) view.findViewById(R.id.contact_name))
                        .setText(contact.getName());
            } else {
                // a contact can have multiple associated numbers
                view = GraphicalOutputUtils.inflate(context,
                        R.layout.skill_telephone_contact_secondary_number);
            }

            final String number = numbers.get(j);
            ((TextView) view.findViewById(R.id.contact_number)).setText(number);
            view.setOnClickListener(v -> ConfirmCallOutput.call(context, number));
            output.addView(view);
        }
    }
}

File: app/src/main/kotlin/org/stypox/dicio/skills/telephone/TelephoneOutput.kt
package org.stypox.dicio.skills.telephone

import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.unit.dp
import org.dicio.skill.skill.Skill
import org.dicio.skill.context.SkillContext
import org.dicio.skill.skill.InteractionPlan
import org.dicio.skill.skill.SkillOutput
import org.stypox.dicio.R
import org.stypox.dicio.io.graphical.Headline
import org.stypox.dicio.util.getString

class TelephoneOutput(
    private val contacts: List<Pair<String, List<String>>>,
) : SkillOutput {
    override fun getSpeechOutput(ctx: SkillContext): String = if (contacts.isEmpty()) {
        ctx.getString(R.string.skill_telephone_unknown_contact)
    } else {
        ctx.getString(R.string.skill_telephone_found_contacts, contacts.size)
    }

    override fun getInteractionPlan(ctx: SkillContext): InteractionPlan {
        val nextSkills = mutableListOf<Skill<*>>(
            ContactChooserName(
                // when saying the name, there is no way to distinguish between
                // different numbers, so just use the first one
                contacts.map { Pair(it.first, it.second[0]) }
            )
        )

        if (ctx.parserFormatter != null) {
            nextSkills.add(
                ContactChooserIndex(
                    contacts.flatMap { contact ->
                        contact.second.map { number ->
                            Pair(contact.first, number)
                        }
                    }
                )
            )
        }

        return InteractionPlan.StartSubInteraction(
            reopenMicrophone = true,
            nextSkills = nextSkills,
        )
    }

    @Composable
    override fun GraphicalOutput(ctx: SkillContext) {
        if (contacts.isEmpty()) {
            Headline(text = getSpeechOutput(ctx))
        } else {
            Column {
                for (i in contacts.indices) {
                    if (i != 0) {
                        Spacer(modifier = Modifier.height(8.dp))
                    }

                    Text(
                        text = contacts[i].first,
                        style = MaterialTheme.typography.titleMedium,
                    )

                    for (number in contacts[i].second) {
                        Text(
                            text = number,
                            style = MaterialTheme.typography.bodyMedium,
                            modifier = Modifier.padding(start = 8.dp, top = 2.dp)
                        )
                    }
                }
            }
        }
    }
}
